pallets/providers/src/tests.rs

```diff
@@ -23,1040 +27,2509 @@ type DepositPerData = <Test as crate::Config>::DepositPerData;
+                        Event::<Test>::SignUpRequestCanceled { who: bob }.into(),
+                    );
+                });
+            }
 
-            // Sign up Alice as a Backup Storage Provider
-            assert_ok!(StorageProviders::bsp_sign_up(
-                RuntimeOrigin::signed(alice),
-                storage_amount,
-                multiaddresses.clone(),
-            ));
-
-            // Try to sign up Alice now as a Main Storage Provider
-            // We use assert_noop to make sure that it not only returns the specific
-            // error, but it also does not modify any storage
-            assert_noop!(
-                StorageProviders::msp_sign_up(
-                    RuntimeOrigin::signed(alice),
-                    storage_amount,
-                    multiaddresses.clone(),
-                    value_prop.clone()
-                ),
-                Error::<Test>::AlreadyRegistered
-            );
+            #[test]
+            fn confirm_sign_up_is_free_if_successful() {
+                ExtBuilder::build().execute_with(|| {
+                    // Initialize variables:
+                    let mut multiaddresses: BoundedVec<
+                        MultiAddress<Test>,
+                        MaxMultiAddressAmount<Test>,
+                    > = BoundedVec::new();
+                    multiaddresses.force_push(
+                        "/ip4/127.0.0.1/udp/1234"
+                            .as_bytes()
+                            .to_vec()
+                            .try_into()
+                            .unwrap(),
+                    );
+                    let value_prop: ValueProposition<Test> = ValueProposition {
+                        identifier: ValuePropId::<Test>::default(),
+                        data_limit: 10,
+                        protocols: BoundedVec::new(),
+                    };
+                    let storage_amount: StorageData<Test> = 100;
+
+                    // Get the Account Id of Alice
+                    let alice: AccountId = 0;
+
+                    // Request sign up of Alice as a Main Storage Provider
+                    assert_ok!(StorageProviders::request_msp_sign_up(
+                        RuntimeOrigin::signed(alice),
+                        storage_amount,
+                        multiaddresses.clone(),
+                        value_prop.clone()
+                    ));
+
+                    // Advance enough blocks for randomness to be valid
+                    run_to_block(
+                        frame_system::Pallet::<Test>::block_number()
+                            + BLOCKS_BEFORE_RANDOMNESS_VALID,
+                    );
+
+                    // Confirm the sign up of the account as a Main Storage Provider
+                    let confirm_result =
+                        StorageProviders::confirm_sign_up(RuntimeOrigin::signed(alice), alice);
+                    assert_eq!(confirm_result, Ok(Pays::No.into()));
```

Comment:
Could think of checking the balances before and after, although I don't know if mocked runtimes consume gas 🤔

Reply:
They don't! It was actually a problem I ran into when developing a free transactions pallet (where the runtime would sponsor transactions for users). I learned then that the easiest way to check if a user would have paid for an extrinsic execution is to check in the returned`DispatchResultWithPostInfo` the `pays` field and making sure it is `Pays::No`